drivers: clock: SAM D5x/E5x: Enhance clock control

sunil-abraham opened the pull request:
This update extends the functionality of clock control driver for the SAM D5x/E5x series. It introduces bindings for clock sources and generators, including DFLL, FDPLL, GCLK generator, MCLK CPU, OSC32K, RTC clock, and XOSC.

The driver now supports bootup clock initialization, asynchronous operation, and a set_rate API for dynamic clock management.
On the board side, the configuration has been updated to set the CPU clock to 120MHz using the FDPLL, sourced from the external XOSC

sunil-abraham committed:
dts: clock: SAM D5x/E5x: add more functionality

Add more functionality in clock control driver.
Add bindings for dfll, fdpll, gclk generator, mclk cpu, osc32k, rtc clock
and xosc.

Signed-off-by: Sunil Abraham <sunil.abraham@microchip.com>

sunil-abraham committed:
drivers: clock_control: SAM D5x/E5x: add more functionality

Add more functionality in clock control driver.
Implement bootup clock initialization, async on and set_rate API.

Signed-off-by: Sunil Abraham <sunil.abraham@microchip.com>

ArunMCHP commented on drivers/clock_control/clock_control_mchp_sam_d5x_e5x.c:
```diff
@@ -115,6 +116,97 @@ typedef union {
+#if CONFIG_CLOCK_CONTROL_MCHP_CONFIG_BOOTUP
+
+/** @brief XOSC initialization structure. */
+typedef struct {
```
Don't typedef enum and struct.

sunil-abraham replied:
Updated.

ArunMCHP commented on drivers/clock_control/clock_control_mchp_sam_d5x_e5x.c:
```diff
@@ -669,13 +1542,924 @@ static int clock_mchp_get_rate(const struct device *dev, clock_control_subsys_t
+	if (src <= CLOCK_MCHP_FDPLL_SRC_GCLK11) {
+		gclk_regs->GCLK_PCHCTRL[inst + 1] |=
+			(GCLK_PCHCTRL_GEN(src) | GCLK_PCHCTRL_CHEN_Msk);
+		while ((gclk_regs->GCLK_PCHCTRL[inst + 1] & GCLK_PCHCTRL_CHEN_Msk) == 0) {
```
Please use the WAIT_FOR macro with a timeout

ArunMCHP commented on dts/bindings/clock/microchip/sam_d5x_e5x/microchip,sam-d5x-e5x-dfll.yaml:
```diff
@@ -0,0 +1,169 @@
+      - 1
+    default: 0
+    description: |
+      Disable chill cycle
```
Please provide more descriptive comment.

sunil-abraham commented on boards/microchip/sam/sam_e54_xpro/sam_e54_xpro.yaml:
```diff
@@ -13,4 +13,5 @@ supported:
   - pinctrl
   - shell
   - uart
+  - clock
```
Updated.

sunil-abraham commented on drivers/clock_control/clock_control_mchp_sam_d5x_e5x.c:
```diff
@@ -669,13 +1542,924 @@ static int clock_mchp_get_rate(const struct device *dev, clock_control_subsys_t
+	val |= ((gclkgen_init->enable != 0) ? GCLK_GENCTRL_GENEN(1) : 0);
+
+	config->gclk_regs->GCLK_GENCTRL[inst] = val;
+	while (config->gclk_regs->GCLK_SYNCBUSY != 0) {
```
Updated.

nandojve commented on dts/arm/microchip/sam/sam_d5x_e5x/common/samd5xe5x.dtsi:
```diff
@@ -42,12 +42,43 @@
 			      <0x40001400 0x20>, <0x40001c00 0x140>;
 			reg-names = "mclk", "oscctrl",
 				    "osc32kctrl", "gclk";
+			interrupts = <1 0>, <2 0>, <3 0>, <4 0>, <5 0>, <6 0>, <7 0>;
```
I recommend add the `interrupt-names` property to have documented, at least, what are all those 8 interrupts.

ArunMCHP commented on drivers/clock_control/clock_control_mchp_sam_d5x_e5x.c:
```diff
@@ -34,13 +34,23 @@ LOG_MODULE_REGISTER(clock_mchp_sam_d5x_e5x, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
 #define FREQ_1KHZ       1024
 #define FREQ_DFLL_48MHZ 48000000
 
+/* timeout values */
```
```suggestion
/* timeout values in micro seconds */
```

nandojve commented:
Hi @sunil-abraham ,

You need to fix the coding guidelines.

sunil-abraham commented on drivers/clock_control/clock_control_mchp_sam_d5x_e5x.c:
```diff
@@ -5,7 +5,7 @@
 
 /**
- * @file clock_mchp_sam_d5x_e5x.c
+ * @file clock_control_mchp_sam_d5x_e5x.c
```
Shall follow the suggested approach for future PRs.

sunil-abraham commented on boards/microchip/sam/sam_e54_xpro/sam_e54_xpro.yaml:
```diff
@@ -10,6 +10,7 @@ toolchain:
 flash: 1024
 ram: 256
 supported:
+  - clock
```
Corrected.

sunil-abraham committed:
boards: microchip: SAM_D5x_E5x: configure clock source and frequency

Configure CPU clock using FDPLL, sourced from XOSC to achieve 120Mhz.
Also configure UART clock to support bootup initialization.

Signed-off-by: Sunil Abraham <sunil.abraham@microchip.com>
